rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+`tcp` type will be used in the app's URI. Additionally, if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI, but annotation supersedes hint from the `appProtocol`.
+
+Annotation `teleport.dev/discovery-port` will control preferred port for the Kubernetes service. Its value should be one of the
+exposed service ports; otherwise, the app will not be imported. Value can be matched either by numeric value or by the name of
+the port defined on the service.
+
+Example Kubernetes service configs using these annotations can look like this:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: first-service
+  labels:
+    app: first-service
+  annotations:
+    teleport.dev/discovery-type: app # Allowed values are [`app`]
+    teleport.dev/discovery-protocol: http # Allowed values are [`http`, `https`, `tcp`]
+    teleport.dev/discovery-port: 80
+spec:
+  ports:
+  - name: http
+    port: 80
+    protocol: TCP
+    targetPort: 80
+  - name: postgres
+    port: 5432
+    protocol: TCP
+    targetPort: 5432
+  selector:
+    app: first-pod
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: second-service
+  labels:
+    app: second-service
+  annotations:
+    teleport.dev/discovery-protocol: tcp # Allowed values are [`http`, `https`, `tcp`]
+    teleport.dev/discovery-port: fluentd
+spec:
+  ports:
+    - name: http
+      port: 80
+      protocol: TCP
+      targetPort: 80
+    - name: fluentd
+      port: 24224
+      protocol: TCP
+      targetPort: 24224
+  selector:
+    app: second-pod
+```
+
+Discovery for the Kubernetes apps will support dynamic discovery configuration, when it will be implemented - [RFD 125](https://github.com/gravitational/teleport/blob/master/rfd/0125-dynamic-auto-discovery-config.md).
+
+### App service
+
+App service will be running inside the Kubernetes cluster and proxy connection from remote users to local cluster IP. App service
+already has capability to watch dynamically created app resources, users just need to correctly setup
+`resources` field in the config:
+
+```yaml
+app_service:
+  enabled: yes
+  resources:
+    - labels:
+        "teleport.dev/kubernetes-cluster": "main-cluster"
+        "env": "staging"
+```
+
+### Helm chart
+
+We will update helm chart `teleport-kube-agent` to support configuring Kubernetes apps discovery. Ability to configure
+and deploy discovery service will be added by using parameter `kubernetesDiscovery`.
+
+```yaml
+kubernetesDiscovery:
+  - types: ["app"]
+    namespaces: [ "toronto", "porto" ]
+    labels:
+      env: staging
+  - types: ["app"]
+    namespaces: [ "seattle", "oakland" ]
+    labels:
+      env: testing
+```
+
+## Security
+
+Similar to the in-cluster kubernetes service, discovery will use service account of its pod as credentials to list services inside the cluster.
```

Comment:
Are we OK with picking services of the `ExternalService` type pointing to resources living outside of the cluster? This sounds like a handy hack to expose an app or a DB living next to a cluster without deploying an additional Teleport agent, but discovering apps out of the cluster might raise additional security considerations.